crates/hotshot-state-prover/src/circuit.rs

```diff
@@ -0,0 +1,123 @@
+//! Circuit implementation for verifying light client state update
+
+use std::marker::PhantomData;
+
+use ark_ec::twisted_edwards::TECurveConfig;
+use ethereum_types::U256;
+use hotshot_types::traits::stake_table::{SnapshotVersion, StakeTableScheme};
+use jf_plonk::errors::PlonkError;
+use jf_primitives::{
+    circuit::signature::schnorr::VerKeyVar,
+    rescue::RescueParameter,
+    signatures::{
+        bls_over_bn254::VerKey as BLSVerKey,
+        schnorr::{Signature, VerKey as SchnorrVerKey},
+    },
+};
+use jf_relation::gadgets::ecc::TEPoint;
+use jf_relation::{
+    errors::CircuitError, gadgets::ecc::SWToTEConParam, BoolVar, Circuit, PlonkCircuit,
+};
+
+use self::{
+    config::{u256_to_field, NUM_ENTRIES},
+    types::{HotShotState, StakeTableEntryVar},
+};
+
+pub mod config;
+pub mod types;
+
+#[derive(Clone, Debug)]
+pub struct StateUpdateBuilder<F: RescueParameter>(PhantomData<F>);
+
+impl<F> StateUpdateBuilder<F>
+where
+    F: RescueParameter + SWToTEConParam,
+{
+    /// A function that takes as input:
+    /// - stake table entries (`Vec<(BLSVerKey, SchnorrVerKey, Amount)>`)
+    /// - schnorr signatures of the updated states (`Vec<SchnorrSignature>`)
+    /// - updated hotshot state (`(view_number, block_height, block_comm, fee_ledger_comm, stake_table_comm)`)
+    /// - signer bit vector
+    /// - quorum threshold
+    /// checks that
+    /// - the signer's accumulated weight exceeds the quorum threshold
+    /// - the commitment of the stake table
+    /// - all schnorr signatures are valid
+    pub fn build<ST, P>(
```

Comment:
this is `fn build_for_preprocessing` instead? (alternatively, maybe via an enum input argument to indicate preprocessing mode or proving mode)

Reply:
I'm thinking whether we should only export a single API for proof generation only, w/o exposing these details about circuit.
These interface issues could be finalized when doing proof generation: #1985